index.d.ts

```diff
@@ -0,0 +1,77 @@
+export interface JsonMLAttributes {
+  [key: string]: any // @TODO - can this be anything but a string? it technically SHOULD be,
+      // but I'm not sure if we want to be that strict. maybe it should be primitives at least?
+      // or maybe we just leave it alone?
+}
+
+export type JsonMLChildNode = string | JsonMLNode;
+
+export type JsonMLChildren = Array<JsonMLChildNode>;
+
+export interface JsonMLNode extends Array<string | JsonMLAttributes | JsonMLChildNode> {
+  0: string;
+}
```

Comment:
My TS skills aren't the best. I wanted to find some way in the type definition to state that attributes could _only_ appear in the second position. But I couldn't find a way to do it.

Reply:
`comment`: It seems a little circular to me that a `ChildNode` can be a `Node` but a `Node` can be an `Array<ChildNode>`. Also `Children` can be an `Array<Node>`. I feel like things could be simplified among these three types but I don't have a lot of context. It could be that this kind of type structure is just what happens when you try to type the DOM!

`question`: As far as the attributes/second position problem, it's not clear to me how this solves it. Is this accomplished by just "reserving" the first position for a string, such that the second position _may_ be an `Attributes`.

Reply:
The issue is that we're trying to represent a tree of nodes as JSON. So, we can't define a schema that has tree edges like usual, necessarily - it _is_ kind of a recursive situation. The children are just more nodes. The only difference is that children may also include text nodes (strings), whereas a text node cannot have children.

The issue of solving the second position problem ... I don't know how to solve it in the definition. Right now, this allows attributes to appear _anywhere_ after the first position. However, It shouldn't be allowed anywhere except the second position. Part of the problem, maybe, is that I'm defining a `JsonMLNode` initially as a union of `string | JsonMLAttributes | JsonMLChildNode` which invariably means all are whitelisted. 
Desired:
```
0: string (only)
1: JsonMLAttrributes | JsonMLNode | string (text node)
2+: JsonMLNode | string (text node)
```
🤷‍♂ 🤷‍♂ 🤷‍♂

Reply:
I think I worked this out finally!

Reply:
`question`: Do you actually need separate types here? Why not something like:
```
interface ElementMaybeWithAttributes extends Array<TagName | Attributes | Element | undefined> {
  0: TagName;
  1?: Attributes;
}
```